### DISCUSSION TITLE
high performance algorithms, low performance graph creation from DataFrame

### DISCUSSION THREAD

**Discussion by clstaudt** [networkit/networkit#958] [Ideas]
Real-world workflows applying graph algorithms often include graph creation from common data structures like the `pandas.DataFrame`. For example, in order to map tabular data to a `networkit.Graph` I wrote this function:

```python
def to_networkit_graph(
    data_copurchase,
    weight_col,
) -> networkit.Graph:
    product_ids = numpy.union1d(data_copurchase["product1"].unique(), data_copurchase["product2"].unique())
    id_product2node = dict(zip(product_ids, range(len(product_ids))))
    id_node2product = {v : k for (k,v) in id_product2node.items()}
    
    G = networkit.Graph(
        directed=False,
        weighted=True,
    )
    G.addNodes(len(product_ids))
    for _, row in tqdm.tqdm(data_copurchase.iterrows()):
        G.addEdge(
            u=id_product2node[row["product1"]],
            v=id_product2node[row["product2"]],
            w=row[weight_col]
        )
    return G, id_product2node, id_node2product
```

Observations:
- this step is relatively slow (12 seconds for 600 000 edges) and will not scale well to large graphs
- building and storing the node id mappings is something I would like the library to handle for me

In practice playing nicely together with the PyData ecosystem data structures (pandas, numpy) is probably one of the most important factors when choosing a graph framework. So I am wondering whether networkit has developed further in this direction so that the function above can become more elegant and performant.

See also:
- https://networkx.org/documentation/networkx-1.10/reference/generated/networkx.convert_matrix.from_pandas_dataframe.html

**Comment by avdgrinten** [comment]
This is something that could definitely be improved. Cython has native support for numpy arrays, so adding a `from_coo_arrays` function or so would be rather simple.

**Comment by fabratu** [comment]
I am converting this to a discussion in order to better track this feature request.
